File: Service/CompanyService.vb
Imports BO
Imports DAL
Imports Facade
Imports System.Linq.Expressions

Public Class CompanyService
    Implements ICompanyService

    Public Function GetCompanyByID(id As Integer) As GetResponse(Of CompanyBO) Implements ICompanyService.GetCompanyByID
        Dim response As New GetResponse(Of CompanyBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyInfoDAO()

        Dim _entity = dao.GetById(id)
        Dim company As New CompanyBO

        Dim err = CompanyTranslator.TranslateEntityToBO(_entity, company)
        If err = ErrorCode.Success Then
            response.Value = company
        Else
            response.AddError(err.ToString())
        End If
        Return response
    End Function
    Public Function GetCompanyForSelectByActivity(actid As Integer) As GetResponse(Of IdNameBO) Implements ICompanyService.GetCompanyForSelectByActivity
        Dim response As New GetResponse(Of IdNameBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyInfoDAO()
        Dim entities = dao.GetNoTracking.Where(Function(m) m.Activity.Any(Function(i) i.ActivityID = actid))
        For Each _entity In entities
            response.AddValue(_entity.GetIdName())
        Next
        response.Values = response.Values.OrderBy(Function(m) m.Display).ToList
        Return response
    End Function

    Public Function GetCompanyBySearchfilter(filter As CompanyFilter) As GetResponse(Of CompanyBO) Implements ICompanyService.GetCompanyBySearchfilter
        Dim response As New GetResponse(Of CompanyBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyInfoDAO()
        Dim entitys = dao.GetNoTracking().Where(CompanyQuery.GetFilterQeury(filter)).OrderBy(Function(m) m.BedrijfsNaam)

        For Each _entity In entitys
            Dim company As New CompanyBO
            Dim err = CompanyTranslator.TranslateEntityToBO(_entity, company)
            If err = ErrorCode.Success Then
                response.AddValue(company)
            Else
                response.AddError(err.ToString())
            End If
        Next
        Return response
    End Function
    Public Function GetCompanyNameById(id As Integer) As String Implements ICompanyService.GetCompanyNameById
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyInfoDAO()
        Dim _entity = dao.GetById(id)
        Return _entity.BedrijfsNaam
    End Function
    Public Function GetCompanyForSearchList(searchterm As String) As GetResponse(Of SelectBO) Implements ICompanyService.GetCompanyForSearchList
        Dim response As New GetResponse(Of SelectBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyInfoDAO()
        Dim entitys = dao.GetNoTracking().Where(CompanyQuery.GetNameQuery(searchterm)).OrderBy(Function(m) m.BedrijfsNaam).Select(Function(m) New SelectBO With {.id = m.CompanyID, .text = m.BedrijfsNaam, .extra = "Company"})
        response.Values = entitys.ToList()
        'For Each _entity In entitys
        '    response.AddValue(_entity.GetIdNameForSearch())
        'Next
        Return response
    End Function



    Public Function InsertUpdate(company As CompanyBO) As Response Implements ICompanyService.InsertUpdate
        Dim response As New Response
        If (String.IsNullOrWhiteSpace(company.Bedrijfsnaam)) Then
            response.AddError("companyname is mandatory")
        End If
        If (Not response.Success) Then Return response

        Dim uow As New UnitOfWork()
        Dim _entity As CompanyInfo = Nothing
        If (company.CompanyId = 0) Then
            _entity = uow.GetCompanyInfoDAO().GetNew()
        Else
            _entity = uow.GetCompanyInfoDAO().GetById(company.CompanyId)
        End If
        If (_entity IsNot Nothing) Then
            Dim err = CompanyTranslator.TranslateBOToEntity(_entity, company, uow)
            If (err <> ErrorCode.Success) Then
                response.AddError(err.ToString())
            End If
        Else
            response.AddError("company not found")
        End If

        response.AddError(uow.SaveChanges())
        Return response
    End Function
    Function Delete(ids As List(Of Integer)) As Response Implements ICompanyService.Delete
        Dim response As New Response()
        Dim uow As New UnitOfWork()

        For Each id In ids
            uow.GetCompanyInfoDAO().DeleteObject(id)
        Next
        response.Messages.AddRange(uow.SaveChanges())

        Return response
    End Function
    Function Delete(bos As List(Of CompanyBO)) As Response Implements ICompanyService.Delete
        Return Delete(bos.Select(Function(s) s.CompanyId).ToList())
    End Function
    Public Function GetCompanyActivities(companyid As Integer) As GetResponse(Of ActivityBO) Implements ICompanyService.GetCompanyActivities
        Dim uow As New UnitOfWork(False)
        Dim response As New GetResponse(Of ActivityBO)
        Dim dao = uow.GetCompanyInfoDAO()
        Dim _entity = dao.GetById(companyid)
        Dim activities As New List(Of ActivityBO)
        For Each act In _entity.Activity
            Dim activity As New ActivityBO
            Dim err = ActivityTranslator.TranslateEntityToBO(act, activity)
            If err = ErrorCode.Success Then
                response.AddValue(activity)
            Else
                response.AddError(err.ToString())
            End If
        Next
        Return response
    End Function

    Public Function AddCompanyActivity(companyid As Integer, activityid As Integer) As Response Implements ICompanyService.AddCompanyActivity
        Dim response As New Response
        Dim uow = New UnitOfWork()
        Dim company = uow.GetCompanyInfoDAO().GetById(companyid)
        If (company Is Nothing) Then
            response.AddError("company not found")
            Return response
        End If
        Dim acitvity = uow.GetActivityDAO().GetById(activityid)
        If (acitvity Is Nothing) Then
            response.AddError("activity not found")
            Return response
        End If
        company.Activity.Add(acitvity)
        response.AddError(uow.SaveChanges())
        Return response
    End Function
    Public Function DeleteCompanyActivity(companyid As Integer, activityid As Integer) As Response Implements ICompanyService.DeleteCompanyActivity
        Dim response As New Response
        Dim uow = New UnitOfWork()
        Dim company = uow.GetCompanyInfoDAO().GetById(companyid)
        If (company Is Nothing) Then
            response.AddError("company not found")
            Return response
        End If
        Dim acitvity = uow.GetActivityDAO().GetById(activityid)
        If (acitvity Is Nothing) Then
            response.AddError("activity not found")
            Return response
        End If
        company.Activity.Remove(acitvity)
        response.AddError(uow.SaveChanges())
        Return response
    End Function

End Class

File: ServiceCore/CompanyService.cs
using BOCore;
using FacadeCore;
using DALCore;
using DALCore.Models;
using DALCore.Query;
using ServiceCore.Translators;
using Microsoft.EntityFrameworkCore;
using System.Linq;

namespace ServiceCore
{
    public class CompanyService : ICompanyService
    {
        private readonly UnitOfWorkCore _uow;

        public CompanyService(UnitOfWorkCore uow)
        {
            _uow = uow;
        }

        public GetResponse<CompanyBO> GetCompanyByID(int id)
        {
            var response = new GetResponse<CompanyBO>();

            var entity = _uow.CompanyInfo.GetNoTracking()
                .Where(m => m.CompanyId == id)
                .Include(m => m.PostCode).ThenInclude(m => m.Country)
                .SingleOrDefault();

            var bo = new CompanyBO();
            var err = CompanyTranslator.TranslateEntityToBO(entity, bo);
            if (err == ErrorCode.Success) response.Value = bo;
            else response.AddError(err.ToString());

            return response;
        }

        public GetResponse<IdNameBO> GetCompanyForSelectByActivity(int actid)
        {
            var response = new GetResponse<IdNameBO>();

            var entities = _uow.CompanyInfo.GetNoTracking()
                .Where(m => m.Activity.Any(i => i.ActivityId == actid));

            foreach (var e in entities)
                response.AddValue(e.GetIdName());

            response.Values = response.Values.OrderBy(m => m.Display).ToList();
            return response;
        }

        public GetResponse<CompanyBO> GetCompanyBySearchfilter(CompanyFilter filter)
        {
            var response = new GetResponse<CompanyBO>();

            var entities = _uow.CompanyInfo.GetNoTracking()
                .Where(CompanyQuery.GetFilterQeury(filter))
                .OrderBy(m => m.BedrijfsNaam);

            foreach (var e in entities)
            {
                var bo = new CompanyBO();
                var err = CompanyTranslator.TranslateEntityToBO(e, bo);
                if (err == ErrorCode.Success) response.AddValue(bo);
                else response.AddError(err.ToString());
            }
            return response;
        }

        public string GetCompanyNameById(int id)
        {
            var e = _uow.CompanyInfo.GetById(id);
            return e?.BedrijfsNaam ?? string.Empty;
        }

        public string GetCompanyNameByContractId(int id)
        {
            var e = _uow.Contracts.GetNoTracking()
                .Where(m => m.Id == id)
                .Include(m => m.Company)
                .SingleOrDefault();

            return e?.Company?.BedrijfsNaam ?? string.Empty;
        }

        public GetResponse<SelectBO> GetCompanyForSearchList(string searchterm)
        {
            var response = new GetResponse<SelectBO>();

            var query = _uow.CompanyInfo.GetNoTracking()
                .Where(CompanyQuery.GetNameQuery(searchterm))
                .OrderBy(m => m.BedrijfsNaam)
                .Select(m => new SelectBO { id = m.CompanyId, text = m.BedrijfsNaam, extra = "Company" });

            response.Values = query.ToList();
            return response;
        }

        public Response InsertUpdate(CompanyBO company)
        {
            var response = new Response();
            if (string.IsNullOrWhiteSpace(company.Bedrijfsnaam))
            {
                response.AddError("companyname is mandatory");
                return response;
            }

            CompanyInfo entity = company.CompanyId == 0
                ? _uow.CompanyInfo.GetNew()
                : _uow.CompanyInfo.GetById(company.CompanyId);

            if (entity == null)
            {
                response.AddError("company not found");
                return response;
            }

            // Let op: pas CompanyTranslator aan zodat hij UnitOfWorkCore accepteert.
            var err = CompanyTranslator.TranslateBOToEntity(entity, company, _uow);
            if (err != ErrorCode.Success)
            {
                response.AddError(err.ToString());
                return response;
            }

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Bedrijf aangepast of toegevoegd", "Bedrijf niet aangepast of toegevoegd");
            return response;
        }

        public Response Delete(List<int> ids)
        {
            var response = new Response();

            foreach (var id in ids)
                _uow.CompanyInfo.DeleteObject(id);

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Record(s) verwijderd", "Geen records verwijderd");
            return response;
        }

        public Response Delete(List<CompanyBO> bos)
            => Delete(bos.Select(s => s.CompanyId).ToList());

        public GetResponse<ActivityBO> GetCompanyActivities(int companyid)
        {
            var response = new GetResponse<ActivityBO>();

            var entity = _uow.CompanyInfo.GetNormal()
                .Include(m => m.Activity)
                .FirstOrDefault(m => m.CompanyId == companyid);

            if (entity == null)
            {
                response.AddError("company not found");
                return response;
            }

            foreach (var act in entity.Activity)
            {
                var bo = new ActivityBO();
                var err = ActivityTranslator.TranslateEntityToBO(act, bo);
                if (err == ErrorCode.Success) response.AddValue(bo);
                else response.AddError(err.ToString());
            }
            return response;
        }

        public Response AddCompanyActivity(int companyid, int activityid)
        {
            var response = new Response();

            var company = _uow.CompanyInfo.GetById(companyid);
            if (company == null)
            {
                response.AddError("company not found");
                return response;
            }

            var activity = _uow.Activities.GetById(activityid);
            if (activity == null)
            {
                response.AddError("activity not found");
                return response;
            }

            company.Activity.Add(activity);

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Activiteit gekoppeld", "Activiteit niet gekoppeld");
            return response;
        }

        public Response DeleteCompanyActivity(int companyid, int activityid)
        {
            var response = new Response();

            var company = _uow.CompanyInfo.GetById(companyid);
            if (company == null)
            {
                response.AddError("company not found");
                return response;
            }

            var activity = _uow.Activities.GetById(activityid);
            if (activity == null)
            {
                response.AddError("activity not found");
                return response;
            }

            company.Activity.Remove(activity);

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Activiteit ontkoppeld", "Activiteit niet ontkoppeld");
            return response;
        }
    }
}
